pkg/dataobj/builder.go

```diff
@@ -349,12 +292,12 @@ func (b *Builder) buildObject() error {
 		return fmt.Errorf("encoding object: %w", err)
 	}
 
-	b.metrics.builtSize.Observe(float64(b.flushBuffer.Len()))
+	b.metrics.builtSize.Observe(float64(output.Len()))
 
 	// We pass context.Background() below to avoid allowing building an object to
 	// time out; timing out on build would discard anything we built and would
 	// cause data loss.
-	dec := encoding.ReaderAtDecoder(bytes.NewReader(b.flushBuffer.Bytes()), int64(b.flushBuffer.Len()))
+	dec := encoding.ReaderAtDecoder(bytes.NewReader(output.Bytes()), int64(output.Len()))
 	return b.metrics.encoding.Observe(context.Background(), dec)
```

Comment:
We'll need to take a little bit of caution here since it's up to the caller to pass an empty buffer:

* The built size should be the difference from the pre-flush length to the post-flush length to account for preexisting data 
* Similarly, we need to ignore any preexisting data in the buffer when constructing the ReaderAtDecoder to get stats from the built object.

Adding preexisting data to our test case may be a good idea here, since I think this is subtle and will be easy to forget about when refactoring.